lottie/src/main/java/com/airbnb/lottie/utils/OffscreenLayer.java

```diff
@@ -0,0 +1,494 @@
+      // doing prior to OffscreenLayer, so keep that behavior to be safe and avoid regressions.
+      return RenderStrategy.SAVE_LAYER;
+    }
+
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q && parentCanvas.isHardwareAccelerated()) {
+      if (op.hasShadow() && Build.VERSION.SDK_INT <= Build.VERSION_CODES.S) {
+        // RenderEffect, which we need for shadows, was only introduced in S. This means that, pre-S,
+        // we have to do renderShadow() on a bitmap, like in the BITMAP case. However, since it's not
+        // possible to draw a RenderNode to a software-rendering canvas, there would be no way to get
+        // our RenderNode contents onto a bitmap where we can renderShadow(). Therefore, fall back to
+        // full bitmap mode.
+        return RenderStrategy.BITMAP;
+      }
+
+      return RenderStrategy.RENDER_NODE;
+    }
+
+    // Slowest path: render to a bitmap, add a shadow manually.
+    return RenderStrategy.BITMAP;
+  }
+
+  private Bitmap allocateBitmap(RectF bounds, Bitmap.Config cfg) {
+    // Add 5% to the width and height before allocating. This avoids repeated
+    // reallocations in the worst cases of a bitmap growing every frame, while
+    // still being relatively speedy to blit and operate on.
+    int width = (int)Math.ceil(bounds.width() * 1.05);
+    int height = (int)Math.ceil(bounds.height() * 1.05);
+    Bitmap bmp = Bitmap.createBitmap(width, height, cfg);
+    return bmp;
+  }
+
+  private void deallocateBitmap(Bitmap bitmap) {
+    bitmap.recycle();
+  }
+
+  private boolean needNewBitmap(@Nullable Bitmap bitmap, RectF bounds) {
+    if (bitmap == null) {
+      return true;
+    }
+
+    if (bounds.width() >= bitmap.getWidth() || bounds.height() >= bitmap.getHeight()) {
+      return true;
+    }
+
+    // If the required area has reduced in size considerably, trigger a reallocation, since
+    // we might be paying a large unnecessary penalty to work with a bitmap that big.
+    if (bounds.width() < bitmap.getWidth() * 0.75f || bounds.height() < bitmap.getHeight() * 0.75f) {
+      return true;
+    }
+
+    return false;
+  }
+
+  public Canvas start(Canvas parentCanvas, RectF bounds, ComposeOp op) {
+    if (this.parentCanvas != null) {
+      throw new RuntimeException("Cannot nest start() calls on a single OffscreenBitmap - call finish() first");
+    }
+
+    // Determine the scaling applied by the parentCanvas' pre-existing transform matrix. This is an optimization
+    // to avoid creating bitmaps (or render nodes) with unreasonable sizes that will get scaled down when drawn
+    // onto parentCanvas anyhow.
+    if (preExistingTransform == null) preExistingTransform = new float[9];
+    parentCanvas.getMatrix().getValues(preExistingTransform);
+
+    float pixelScaleX = preExistingTransform[Matrix.MSCALE_X];
+    float pixelScaleY = preExistingTransform[Matrix.MSCALE_Y];
+
+    RectF scaledBounds = new RectF(
```

Comment:
Similar to above, could this be updated to not do any allocations in the hot path?

Reply:
Done!

But I would sincerely hope that the runtime does some basic escape analysis and can decide to allocate objects like this on the stack. And even if not, I'd assume the GC pressure is a much smaller problem than the actual time we're taking to render the frame.

I wouldn't mention this normally, but there are places in the code where an allocation cannot be avoided (or it would be cumbersome), such as when you want to pass around a complex object. And the places where I did convert local variables to members to avoid allocations are much harder to read now, there are new nullability issues that need to be addressed, etc.

The optimist in me wants to say that the runtime *surely* has a way of dealing with this, given that Java doesn't really have anything like a `struct`, and this is basically the only option.

My memory is foggy, but even looking at major perf-sensitive Java codebases like Minecraft, I recall they don't shy away from per-frame allocations.

You probably have a much stronger intuition on this than me, so let me know your thoughts.